---
title: "Multiomics Analysis Report"
author: "Aymen Maqsood"
date: "2025-01-12"
output:
  html_document:
    df_print: paged
  pdf_document: default
  html_notebook: default
---

# Computational Analysis of Single-Cell RNA and ATAC Sequencing Data

## Introduction

This analysis investigates the molecular underpinnings of aging in hematopoietic stem cells (HSCs) by examining changes in chromatin accessibility and gene expression between young and aged populations. The dataset was sourced from the GEO database (accession GSE57236) and processed using the R packages *Seurat* and *Signac*. Our objective is to elucidate how age-related alterations in the transcriptional and chromatin accessibility landscapes influence HSC function, potentially shedding light on mechanisms of stem cell decline.

### Chromatin Accessibility and Gene Expression Correlation

Chromatin accessibility governs gene regulation by facilitating transcription factor binding and RNA polymerase recruitment. By integrating ATAC-seq (chromatin accessibility) and RNA-seq (gene expression) data, we aim to uncover how structural changes in chromatin correlate with transcriptional activity, offering insights into the regulatory dynamics of aging HSCs.

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  warning = FALSE,
  message = FALSE
)
```

## Methodology
### Loading Libraries


```{r}
library(Signac)
library(Seurat)
```


### Data Source
The dataset, accessible via GEO (GSE190424), comprises high-resolution single-cell ATAC-seq and RNA-seq data from young and aged HSCs. This analysis builds on the research by Itokawa et al. (2022) in Nature Communications ("Epigenetic traits inscribed in chromatin accessibility in aged hematopoietic stem cells," DOI: 10.1038/s41467-022-30374-9, PMID: 35577813). The ATAC-seq peak count matrix and metadata were imported using the *Read10X_h5* function.


```{r}
counts <- Read10X_h5("~/Desktop/Projects/V1.0/single_cell/data/GSM5723631_Young_HSC_filtered_peak_bc_matrix.h5")
meta <- read.csv("~/Desktop/Projects/V1.0/single_cell/data/GSM5723631_Young_HSC_singlecell.csv.gz", header = TRUE, row.names = 1)
```

###  Preprocessing ATAC-seq Data

A *ChromatinAssay* object was created to process the ATAC-seq data, filtering for peaks present in at least 10 cells and cells with a minimum of 200 peaks.

```{r}
chrom_assay <- CreateChromatinAssay(
  counts = counts,
  sep = c(":", "-"),
  genome = "mm10",
  fragments = "~/Desktop/Projects/V1.0/single_cell/data/GSM5723631_Young_HSC_fragments.tsv.gz",
  min.cells = 10,
  min.features = 200
)

data <- CreateSeuratObject(counts = chrom_assay, assay = "peaks", meta.data = meta)
```


### Genome Annotation
Genomic annotations were added using the EnsDb.Mmusculus.v79 package to map peaks to genes and regulatory elements.

```{r}
library(GenomeInfoDb)
library(EnsDb.Mmusculus.v79)
annotations <- GetGRangesFromEnsDb(ensdb = EnsDb.Mmusculus.v79)
seqlevelsStyle(annotations) <- "UCSC"
Annotation(data) <- annotations
```


### Quality Control

I computed several quality metrics to assess data fidelity:

Nucleosome Signal: Ratio of mononucleosome to nucleosome-free fragments.
TSS Enrichment: Enrichment of reads at transcription start sites.
Blacklist Ratio: Proportion of reads in blacklisted regions.
Percentage Reads in Peaks: Fraction of reads mapping to peak regions.

```{r}
data <- NucleosomeSignal(object = data)
data <- TSSEnrichment(object = data, fast = FALSE)
data$blacklist_ratio <- data$blacklist_region_fragments / data$peak_region_fragments
data$pct_reads_in_peaks <- data$peak_region_fragments / data$passed_filters * 100
```


### Violin plot to visualise qc 

```{r}
VlnPlot(
  object = data,
  features = c("peak_region_fragments", "pct_reads_in_peaks", "blacklist_ratio", "nucleosome_signal", "TSS.enrichment"),
  pt.size = 0.1,
  ncol = 5
)
```

Visual inspection via violin plots revealed data quality distributions


- Cells were filtered based on stringent quality thresholds to exclude low-fidelity data:


#### Subsettign data to exclude low-fidelity data

```{r}
data <- subset(
  x = data,
  subset = peak_region_fragments > 3000 &
           peak_region_fragments < 20000 &
           pct_reads_in_peaks > 15 &
           blacklist_ratio < 0.05 &
           nucleosome_signal < 4 &
           TSS.enrichment > 2
)
```

*Quantile-based thresholds to enhance data robustness*

```{r}
low_prf <- quantile(data[["peak_region_fragments"]]$peak_region_fragments, probs = 0.02)
hig_prf <- quantile(data[["peak_region_fragments"]]$peak_region_fragments, probs = 0.98)
low_prp <- quantile(data[["pct_reads_in_peaks"]]$pct_reads_in_peaks, probs = 0.02)
high_blr <- quantile(data[["blacklist_ratio"]]$blacklist_ratio, probs = 0.98)
hig_ns <- quantile(data[["nucleosome_signal"]]$nucleosome_signal, probs = 0.98)
low_ts <- quantile(data[["TSS.enrichment"]]$TSS.enrichment, probs = 0.02)

data <- subset(
  x = data,
  subset = peak_region_fragments > low_prf &
           peak_region_fragments < hig_prf &
           pct_reads_in_peaks > low_prp &
           blacklist_ratio < high_blr &
           nucleosome_signal < hig_ns &
           TSS.enrichment > low_ts
)
```



### Normalisation and Dimensionality Reduction
Data were normalized using *TF-IDF*, followed by singular value decomposition (*SVD*) for dimensionality reduction, and *UMAP* for visualization. Clustering was performed using a ***graph-based approach***.
```{r}
data <- RunTFIDF(data)
data <- FindTopFeatures(data, min.cutoff = "q0")
data <- RunSVD(data)
data <- RunUMAP(object = data, reduction = "lsi", dims = 2:30)
data <- FindNeighbors(object = data, reduction = "lsi", dims = 2:30)
data <- FindClusters(object = data, verbose = FALSE, algorithm = 3)
DimPlot(object = data, label = TRUE) + NoLegend()
```


### Merging Young and Aged HSC Data

- A custom function, import_atac, was defined to import and preprocess ATAC-seq data from young and aged HSCs, followed by merging the datasets.

```{r}
import_atac <- function(count_path, meta_path, fragment_path) {
  counts <- Read10X_h5(filename = count_path)
  meta <- read.csv(file = meta_path, header = TRUE, row.names = 1)
  chrom_assay <- CreateChromatinAssay(
    counts = counts,
    sep = c(":", "-"),
    genome = "mm10",
    fragments = fragment_path,
    min.cells = 10,
    min.features = 200
  )
  data <- CreateSeuratObject(counts = chrom_assay, assay = "peaks", meta.data = meta)
  Annotation(data) <- annotations
  data <- NucleosomeSignal(object = data)
  data <- TSSEnrichment(object = data, fast = FALSE)
  data$blacklist_ratio <- data$blacklist_region_fragments / data$peak_region_fragments
  data$pct_reads_in_peaks <- data$peak_region_fragments / data$passed_filters * 100

  low_prf <- quantile(data[["peak_region_fragments"]]$peak_region_fragments, probs = 0.02)
  hig_prf <- quantile(data[["peak_region_fragments"]]$peak_region_fragments, probs = 0.98)
  low_prp <- quantile(data[["pct_reads_in_peaks"]]$pct_reads_in_peaks, probs = 0.02)
  high_blr <- quantile(data[["blacklist_ratio"]]$blacklist_ratio, probs = 0.98)
  hig_ns <- quantile(data[["nucleosome_signal"]]$nucleosome_signal, probs = 0.98)
  low_ts <- quantile(data[["TSS.enrichment"]]$TSS.enrichment, probs = 0.02)

  data <- subset(
    x = data,
    subset = peak_region_fragments > low_prf &
             peak_region_fragments < hig_prf &
             pct_reads_in_peaks > low_prp &
             blacklist_ratio < high_blr &
             nucleosome_signal < hig_ns &
             TSS.enrichment > low_ts
  )
  return(data)
}
```


### Importing and Merging Data

```{r}
young <- import_atac(
  "~/Desktop/Projects/V1.0/single_cell/data/GSM5723631_Young_HSC_filtered_peak_bc_matrix.h5",
  "~/Desktop/Projects/V1.0/single_cell/data/GSM5723631_Young_HSC_singlecell.csv.gz",
  "~/Desktop/Projects/V1.0/single_cell/data/GSM5723631_Young_HSC_fragments.tsv.gz"
)

old <- import_atac(
  "~/Desktop/Projects/V1.0/single_cell/data/GSM5723632_Aged_HSC_filtered_peak_bc_matrix.h5",
  "~/Desktop/Projects/V1.0/single_cell/data/GSM5723632_Aged_HSC_singlecell.csv.gz",
  "~/Desktop/Projects/V1.0/single_cell/data/GSM5723632_Aged_HSC_fragments.tsv.gz"
)
```

### Merging Young and Aged HSC Data
```{r}
young$dataset <- "young"
old$dataset <- "old"
data <- merge(young, old)
```

- Post-merging, the dataset underwent additional normalization and clustering:

```{r}
data <- FindTopFeatures(data, min.cutoff = "q0")
data <- RunTFIDF(data)
data <- RunSVD(data)
data <- RunUMAP(object = data, reduction = "lsi", dims = 2:30)
data <- FindNeighbors(object = data, reduction = "lsi", dims = 2:30)
data <- FindClusters(object = data, verbose = FALSE, algorithm = 3, resolution = 0.4)
DimPlot(object = data, label = TRUE, group.by = "dataset") + NoLegend()

```

## Data Analysis

### Gene Activity integration

Gene activity scores, derived from ATAC-seq peaks, were computed as a proxy for transcriptional potential and integrated as an RNA assay.

```{r}
gene.activities <- GeneActivity(data)
data[["RNA"]] <- CreateAssayObject(counts = gene.activities)
data <- NormalizeData(
  object = data,
  assay = "RNA",
  normalization.method = "LogNormalize",
  scale.factor = median(data$nCount_RNA)
)

DefaultAssay(data) <- "RNA"
FeaturePlot(
  object = data,
  features = c("Kit", "Pecam1", "Itgam"),
  max.cutoff = "q95"
)
```

### Differential Accessibility Analysis

Differentially accessible regions (DARs) between young and aged HSCs were identified using a logistic regression test, with peak region fragments as a latent variable.

```{r}
DefaultAssay(data) <- "peaks"
da_peaks <- FindMarkers(
  object = data,
  ident.1 = rownames(data[[]][data$dataset == "old", ]),
  ident.2 = rownames(data[[]][data$dataset == "young", ]),
  min.pct = 0.05,
  test.use = "LR",
  latent.vars = "peak_region_fragments"
)
da_peaks$closest_gene <- ClosestFeature(data, regions = rownames(da_peaks))$gene_name
da_peaks$distance <- ClosestFeature(data, regions = rownames(da_peaks))$distance
```


### Visualisation
Chromatin accessibility profiles and gene expression patterns were visualized to contextualize DARs:
```{r}
CoveragePlot(
  object = data,
  region = rownames(da_peaks)[2],
  extend.upstream = 10000,
  extend.downstream = 5000,
  group.by = "dataset"
)
```



```{r}
plot1 <- VlnPlot(object = data, features = rownames(da_peaks)[2], group.by = "dataset")
plot2 <- FeaturePlot(object = data, features = rownames(da_peaks)[2], max.cutoff = "q95")
plot1 | plot2



```


## Results
1) Distinct Chromatin States: UMAP projections of merged data demonstrate robust separation between young and aged HSCs, reflecting distinct chromatin accessibility profiles indicative of age-related epigenetic remodeling.

2)Regulatory Alterations: DAR analysis pinpointed specific genomic regions with altered accessibility in aged HSCs. These regions, often proximal to genes involved in stem cell maintenance (e.g., Kit), suggest regulatory shifts impacting HSC self-renewal and differentiation capacity.

3)Gene Expression Correlation: Integration of gene activity scores with ATAC-seq data revealed that changes in chromatin accessibility strongly correlate with expression levels of key HSC markers (Kit, Pecam1, Itgam). For instance, reduced accessibility near Kit in aged HSCs aligns with diminished stemness, a hallmark of aging.

3)Biological Implications: The proximity of DARs to genes critical for HSC function (e.g., within 10 kb of transcription start sites) implies that chromatin restructuring drives transcriptional reprogramming, potentially contributing to the functional decline observed in aged HSCs.

## Summary 

Summary
- Clustering Patterns: UMAP visualizations delineate clear epigenetic divergence between young and aged HSC populations, underscoring the impact of aging on chromatin architecture.

- Chromatin Dynamics: DARs highlight age-dependent regulatory changes, with enriched accessibility near genes linked to inflammation and reduced accessibility near stemness genes in aged HSCs.

- Candidate Genes: Integration of multiomics data identifies Kit, Pecam1, and Itgam as potential mediators of phenotypic differences, warranting further functional validation.

## Conlclusion:

The distinct chromatin accessibility and gene expression profiles between young and aged HSCs reveal a molecular basis for age-related functional decline. These findings enhance our understanding of epigenetic regulation in stem cell aging and suggest potential targets for therapeutic intervention to mitigate HSC deterioration.